File: src/main/java/spring/webmvc/domain/model/entity/Order.java
package spring.webmvc.domain.model.entity;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import spring.webmvc.domain.model.enums.OrderStatus;
import spring.webmvc.infrastructure.exception.InvalidEntityStatusException;

@Entity
@Table(name = "orders")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Order extends BaseTime {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	private Instant orderedAt;

	@Enumerated(EnumType.STRING)
	private OrderStatus status;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "user_id")
	private User user;

	@OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
	private final List<OrderProduct> orderProducts = new ArrayList<>();

	public List<OrderProduct> getOrderProducts() {
		return Collections.unmodifiableList(orderProducts);
	}

	public static Order create(User user) {
		Order order = new Order();

		order.orderedAt = Instant.now();
		order.status = OrderStatus.ORDER;
		order.user = user;

		return order;
	}

	public void addProduct(Product product, Long quantity) {
		OrderProduct orderProduct = OrderProduct.create(this, product, quantity);

		orderProducts.add(orderProduct);
	}

	public void cancel() {
		if (status == OrderStatus.CONFIRM) {
			throw new InvalidEntityStatusException(
				Order.class,
				Objects.requireNonNull(id),
				status.name(),
				OrderStatus.CANCEL.name()
			);
		}

		status = OrderStatus.CANCEL;

		orderProducts.forEach(OrderProduct::cancel);
	}

	public void updateStatus(OrderStatus status) {
		if (this.status == OrderStatus.CONFIRM) {
			throw new InvalidEntityStatusException(
				Order.class,
				Objects.requireNonNull(id),
				this.status.name(),
				status.name()
			);
		}

		this.status = status;

		if (status == OrderStatus.CANCEL) {
			orderProducts.forEach(OrderProduct::cancel);
		}
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/model/entity/Order.kt
package spring.web.kotlin.domain.model.entity

import jakarta.persistence.*
import org.springframework.http.HttpStatus
import spring.web.kotlin.domain.model.enums.DeliveryStatus
import spring.web.kotlin.domain.model.enums.OrderStatus
import spring.web.kotlin.presentation.exception.BaseException
import spring.web.kotlin.presentation.exception.ErrorCode
import java.time.LocalDateTime

@Entity
@Table(name = "orders")
class Order protected constructor(
    @Id
    @GeneratedValue
    @Column(name = "order_id")
    val id: Long? = null,

    val orderDate: LocalDateTime,

    @Enumerated(EnumType.STRING)
    var status: OrderStatus,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    val member: Member,

    @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    @JoinColumn(name = "delivery_id")
    var delivery: Delivery? = null,

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    private val _orderItems: MutableList<OrderItem> = mutableListOf(),
) : Base() {
    @get:Transient
    val orderItems: List<OrderItem>
        get() = _orderItems.toList()

    companion object {
        fun create(member: Member, delivery: Delivery, orderItems: List<OrderItem>) =
            Order(
                member = member,
                status = OrderStatus.ORDER,
                orderDate = LocalDateTime.now()
            ).apply {
                associateDelivery(delivery)
                orderItems.forEach { associateItem(it) }
            }
    }

    fun associateDelivery(delivery: Delivery) {
        this.delivery = delivery
        delivery.order = this
    }

    fun associateItem(orderItem: OrderItem) {
        _orderItems.add(orderItem)
        orderItem.order = this
    }

    fun cancel() {
        if (delivery?.status == DeliveryStatus.COMP) {
            throw BaseException(ErrorCode.ORDER_CANCEL_NOT_ALLOWED, HttpStatus.UNPROCESSABLE_ENTITY)
        }

        status = OrderStatus.CANCEL
        orderItems.forEach { it.cancel() }
        delivery?.cancel()
    }
}
